lib/connectors_async/job.rb

```diff
@@ -0,0 +1,88 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+require 'connectors_shared/job_status'
+
+# The class is actually supported for single-threaded usage EXCEPT for :documents field
+# :documents are a Queue that's stated to be safe in a threaded environment
+module ConnectorsAsync
+  class Job
+    class StatusUpdateError < StandardError; end
+    def initialize(job_id)
+      @data = {
+        :job_id => job_id,
+        :status => ConnectorsShared::JobStatus::CREATED,
+        :documents => Queue.new
+      }
+    end
+
+    def id
+      @data[:job_id]
+    end
+
+    def status
+      @data[:status]
+    end
+
+    def is_failed?
+      @data[:status] == ConnectorsShared::JobStatus::FAILED
+    end
+
+    def error
+      @data[:error]
+    end
+
+    def has_cursors?
+      @data.has_key?(:cursors)
+    end
+
+    def cursors
+      @data[:cursors]
+    end
+
+    def update_status(new_status)
```

Comment:
should validate `new_status` is a valid?

Reply:
Do you mean check that the variable value is valid or check that the transition to new status is possible?

Reply:
Yes, we can't blindly trust the data from outside world 😄

Reply:
I've added the validation in 9a70a314d64a9a80bffa92e78442849e02b4ab7c

It brings me to a future discussion about - how much input to the methods do we validate? E.g. do we validate for numbers if we take an argument that is treated like a number? Same with strings? How far should we go with this validation?